implementations/rust/ockam/ockam_transport_udp/src/workers/sender.rs

```diff
@@ -38,24 +43,54 @@ impl Worker for UdpSendWorker {
 
     async fn handle_message(
         &mut self,
-        ctx: &mut Context,
+        _ctx: &mut Context,
         msg: Routed<Self::Message>,
     ) -> Result<()> {
-        let mut msg = LocalMessage::decode(msg.payload())?.into_transport_message();
-
-        // Remove sender address
+        // Parse message and remove our address from its routing
+        let mut msg = msg.into_transport_message();
         msg.onward_route.step()?;
 
-        let (peer_addr, _) = match String::from_utf8(msg.onward_route.step()?.deref().clone()) {
-            Ok(s) => UdpRouterHandle::resolve_peer(s)?,
-            Err(_e) => return Err(TransportError::UnknownRoute.into()),
-        };
+        trace!("Sending message to {:?}", msg.onward_route);
+
+        // Resolve peer address to IPv4 SocketAddr(s).
+        let peer_addr = msg.onward_route.step()?;
 
-        if self.sink.send((msg, peer_addr)).await.is_err() {
-            warn!("Failed to send message to peer {}", peer_addr);
-            ctx.stop_worker(ctx.address()).await?;
+        if peer_addr.transport_type() != UDP {
+            error!(addr = %peer_addr,
+                "Destination address is not UDP");
+            return Err(TransportError::UnknownRoute.into());
         }
 
-        Ok(())
+        let peer_addr = peer_addr.address();
+        let peer_addrs = peer_addr
+            .to_socket_addrs()
+            .map_err(|_| TransportError::InvalidAddress)?;
+        let peer_addrs: Vec<_> = peer_addrs.filter(SocketAddr::is_ipv4).collect();
+
+        // Try to send to first SocketAddr
+        let addr = match peer_addrs.first() {
+            Some(a) => *a,
+            None => {
+                warn!("No IPv4 address resolved for peer {:?}", peer_addr);
+                return Err(TransportError::UnknownRoute.into());
+            }
+        };
+        match self.sink.send((msg.clone(), addr)).await {
+            Ok(()) => {
+                trace!("Successful send to {}", addr);
+                Ok(())
+            }
+            Err(e) => {
+                warn!("Failed send to {}: {:?}", addr, e);
+
+                // Inform router of our error
+                let _ = self
+                    .router_handle
+                    .sender_listener_error(self.local_addr)
```

Comment:
Same as before, can we avoid breaking it?

Reply:
See related comments on `listener.rs`.